src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/AttachedCollections/IVsHierarchyItemExtensions.cs

```diff
@@ -26,36 +23,54 @@ internal static class IVsHierarchyItemExtensions
         /// <returns><see langword="true"/> if the target was found, otherwise <see langword="false"/>.</returns>
         public static bool TryFindTarget(this IVsHierarchyItem item, [NotNullWhen(returnValue: true)] out string? target)
         {
-            s_targetFlagsRegex ??= new Regex(@"^(?=.*\b" + nameof(Flags.TargetNode) + @"\b)(?=.*\$TFM:(?<target>[^ ]+)\b).*$", RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.CultureInvariant);
-
             for (IVsHierarchyItem? parent = item; parent is not null; parent = parent.Parent)
             {
-                if (parent.TryGetFlagsString(out string? flagsString))
+                if (parent.TryGetFlags(out ProjectTreeFlags flags) && flags.Contains(Flags.TargetNode))
                 {
-                    Match match = s_targetFlagsRegex.Match(flagsString);
-                    if (match.Success)
+                    // Found an ancestor target node.
+                    const string prefix = "$TFM:";
+                    string? flag = flags.FirstOrDefault(f => f.StartsWith(prefix));
+
+                    if (flag is not null)
                     {
-                        target = match.Groups["target"].Value;
+                        target = flag.Substring(prefix.Length);
                         return true;
                     }
+
+                    // Target node didn't have a TFM flag for some reason. This is unexpected, but there's not much
+                    // we can do about it. Don't check any more ancestors, as there should only be one relevant target node.
+                    break;
                 }
             }
 
             target = null;
             return false;
         }
 
-        public static bool TryGetFlagsString(this IVsHierarchyItem item, [NotNullWhen(returnValue: true)] out string? flagsString)
+        public static bool TryGetFlags(this IVsHierarchyItem item, out ProjectTreeFlags flags)
         {
-            IVsHierarchyItemIdentity identity = item.HierarchyIdentity;
+            // Browse objects are created lazily, so access the one on the project root in order to obtain the unconfigured project.
+            IVsHierarchyItem root = item;
+            while (!root.HierarchyIdentity.IsRoot)
+            {
+                root = root.Parent;
+            }
 
-            if (identity.NestedHierarchy.GetProperty(identity.NestedItemID, (int)__VSHPROPID7.VSHPROPID_ProjectTreeCapabilities, out object? value) == HResult.OK)
+            if (root.HierarchyIdentity.NestedHierarchy is IVsBrowseObjectContext context)
             {
-                flagsString = (string)value;
-                return true;
+                IProjectTreeServiceState? tree = context.UnconfiguredProject.Services.ProjectTreeService?.CurrentTree;
+
+                if (tree is not null)
+                {
+                    if (tree.Tree.TryFind((IntPtr)item.HierarchyIdentity.ItemID, out IProjectTree? subtree))
```

Comment:
when ItemID == VSConstants.VSITEMID.Root, it is the root of the tree. That is the only node where the ItemId doesn't match the id in the tree, so this logic would fail on that node. On the other hand, we don't attach dependencies node under the project node directly, so it would not have any product impact.

Reply:
Thanks. I will add a check to ensure we don't pass `VSConstants.VSITEMID.Root`. We are called for all nodes in the tree during, so will definitely be called for the project node, though we will never want to attach children there so it's fine to skip that node.